marlowe/specification/ReadMe.md

```diff
@@ -0,0 +1,346 @@
+  | MerkleizedInput InputContent BuiltinByteString Contract
+```
+In the above, the `BuiltinByteString` is the hash of the serialized continuation `Contract`.
+
+
+## Plutus Validator for Marlowe Semantics
+
+The Marlowe validator is an unparameterized interpreter of Marlowe semantics. Thus, the script address of the Marlowe validator is independent of the particular contract and roles currency.
+```haskell
+smallMarloweValidator :: MarloweData -> MarloweInput -> ScriptContext -> Bool
+```
+
+
+### Types
+
+The Plutus `Datum` for Marlowe simply bundles the role-currency and payout information with the `State` and the `Contract`:
+```haskell
+data MarloweData =
+    MarloweData
+    {
+      marloweParams   :: MarloweParams
+    , marloweState    :: State
+    , marloweContract :: Contract
+    }
+
+data MarloweParams =
+    MarloweParams
+    {
+      rolePayoutValidatorHash :: ValidatorHash
+    , rolesCurrency           :: CurrencySymbol
+    }
+```
+
+The Plutus `Redeemer` provides the inputs for the semantics:
+```haskell
+type MarloweInput = [MarloweTxInput]
+
+data MarloweTxInput =
+    Input InputContent
+  | MerkleizedTxInput InputContent BuiltinByteString
+```
+In the above, the `BuiltinByteString` is the hash of the serialized continuation of the contract. If `MerkleizedTxInput` is supplied in a redeemer, then the `ScriptContext` for the transaction must also contain an extra entry in its `txInfoData . scriptContextTxInfo` map from `DatumHash` to `Datum` for the serialized continuation of the contract.
+
+
+### Relationship between Marlowe Validator and Semantics
+
+The arguments are `computeTransaction` must be constructed as follows:
+1. The `txInterval` of `TransactionInput` is derived from the `txInfoValidTimeRange . scriptContextInfo` of the `ScriptContext`, as detailed below.
+2. The `txInputs` of `TransactionInput` is derived from the `MarloweInput` provided as the `Redeemer` and the `txInfoData . scriptContextTxInfo` of the `ScriptContext`, as detailed below.
+3. The `State` is the `marloweState` of the `MarloweData` provided as the `Datum`.
+4. The `Contract` is the `marloweContract` of the `MarloweData` provided as the `Datum`.
+5. The new `Datum` at the script address is the `MarloweData` with the same `marloweParams` as originally, but with the new `txOutState` and `txOutContract` of the `TransactionOutput`.
+
+![Relationship between Marlowe validator and semantics.](semantics2plutus.svg)
+
+
+### Specification
+
+Consider the application of the Marlowe validator and Marlowe semantics:
+```haskell
+validationResult = smallMarloweValidator marloweData marloweInput scriptContext
+
+transactionOutput = computeTransaction transactionInput inState inContract
+```
+
+Furthermore, let `marloweValidatorHash :: ValidatorHash` be the hash of the Marlowe semantics script.
+
+The validation fails (via returning `False` for `validationResult` or via the throwing of an error) if any of the following constraints does not hold.
```

Comment:
Should the two forms of failure be explained further? Under what circumstances can an error be thrown?

Reply:
Fundamentally, all failures in Plutus are calls to `error` and all successes are returning `()`. The typed Plutus validators obscure this by using `Bool` as the return type.

I'll try to clean up the discussion in this section.

Reply:
fixed in 8be04323e4ec24c4e923ed214094e85c916ab8e0

Reply:
fixed in 8be04323e4ec24c4e923ed214094e85c916ab8e0